rust/src/shards/gui/widgets/text_input.rs

```diff
@@ -0,0 +1,368 @@
+        0usize
+      } else {
+        self.byte_index_from_char_index(char_index)
+      }
+    } else {
+      0usize
+    };
+
+    if let VarTextBuffer::Editable(var) = self {
+      let text_len = text.len();
+      let (current_len, current_cap) = unsafe {
+        (
+          var.payload.__bindgen_anon_1.__bindgen_anon_2.stringLen as usize,
+          var.payload.__bindgen_anon_1.__bindgen_anon_2.stringCapacity as usize,
+        )
+      };
+
+      if current_cap == 0usize {
+        // new string
+        let tmp = Var::ephemeral_string(text);
+        cloneVar(var, &tmp);
+      } else if (current_cap - current_len) >= text_len {
+        // text can fit within existing capacity
+        unsafe {
+          let base_ptr =
+            var.payload.__bindgen_anon_1.__bindgen_anon_2.stringValue as *mut std::os::raw::c_char;
+          // move the rest of the string to the end
+          std::ptr::copy(
+            base_ptr.add(byte_idx),
+            base_ptr.add(byte_idx).add(text_len),
+            current_len - byte_idx,
+          );
+          // insert the new text
+          let bytes = text.as_ptr() as *const std::os::raw::c_char;
+          std::ptr::copy_nonoverlapping(bytes, base_ptr.add(byte_idx), text_len);
+          // update the length
+          let new_len = current_len + text_len;
+          var.payload.__bindgen_anon_1.__bindgen_anon_2.stringLen = new_len as u32;
+          // fixup null-terminator
+          *base_ptr.add(new_len) = 0;
+        }
+      } else {
+        let mut str = String::from(VarTextBuffer::as_str_from_mut(var));
+        str.insert_str(byte_idx, text);
+        let tmp = Var::ephemeral_string(str.as_str());
+        cloneVar(var, &tmp);
+      }
+
+      text.chars().count()
+    } else {
+      0usize
+    }
+  }
+
+  fn delete_char_range(&mut self, char_range: std::ops::Range<usize>) {
+    assert!(char_range.start <= char_range.end);
+
+    let byte_start = self.byte_index_from_char_index(char_range.start);
+    let byte_end = self.byte_index_from_char_index(char_range.end);
+
+    if byte_start == byte_end {
+      // nothing to do
+      return;
+    }
+
+    if let VarTextBuffer::Editable(var) = self {
+      unsafe {
+        let current_len = var.payload.__bindgen_anon_1.__bindgen_anon_2.stringLen as usize;
+        let base_ptr =
+          var.payload.__bindgen_anon_1.__bindgen_anon_2.stringValue as *mut std::os::raw::c_char;
+        // move rest of the text at the deletion location
+        std::ptr::copy(
+          base_ptr.add(byte_end),
+          base_ptr.add(byte_start),
+          current_len - byte_end,
+        );
+        // update the length
+        let new_len = current_len - byte_end + byte_start;
+        var.payload.__bindgen_anon_1.__bindgen_anon_2.stringLen = new_len as u32;
+        // fixup null-terminator
+        *base_ptr.add(new_len) = 0;
+      }
+    }
+  }
+}
+
+impl VarTextBuffer<'_> {
+  fn as_str_from_mut(var: &mut Var) -> &str {
```

Comment:
Since it's seems exactly the same I seem to be able to replace it with
```rs
fn as_str_from_mut(var: &mut Var) -> &str {
    VarTextBuffer::as_str(var)
  }
```

Reply:
Maybe even call it directly with `&mut Var` but maybe it needs a function in between